provider/provider.go

```diff
@@ -22,8 +24,40 @@ type Provider interface {
 
 // BaseProvider should be inherited by providers
 type BaseProvider struct {
-	Watch    bool
-	Filename string
+	Watch       bool
+	Filename    string
+	Constraints []*types.Constraint
+}
+
+// MatchConstraints must match with EVERY single contraint
+func (p *BaseProvider) MatchConstraints(tags []string, constraints []*types.Constraint) (bool, error) {
+	// if there is no tags and no contraints, filtering is disabled
+	if len(tags) == 0 && len(constraints) == 0 {
+		return true, nil
+	}
+
+	for _, constraint := range constraints {
+		if ok, err := p.matchConstraintWithAtLeastOneTag(tags, constraint); err != nil || xor(ok == true, constraint.MustMatch == true) {
+			return false, err
+		}
+	}
+
+	// If no constraint or every constraints matching
+	return true, nil
+}
+
+func (p *BaseProvider) matchConstraintWithAtLeastOneTag(tags []string, constraint *types.Constraint) (bool, error) {
+	// At the moment, it only supports tags
+	if constraint.Key != "tag" {
+		return false, errors.New("Constraint must be tag-based. Syntax: tag==us-*")
```

Comment:
I would verify this already in `NewConstraint`, I also think this whole function should actually be method of `Constraint` because that's the only instance data it operates on